File: ports/ruby/lib/unicode_ruby/security/identity/emoji_zwj_integrity.rb
# frozen_string_literal: true

require "set"

module UnicodeRuby
  module Security
    module Identity
      # EmojiZwjIntegrity — detection of malformed / unsanctioned emoji
      # ZWJ-sequence shapes per UTS #51 (the identity-layer detector I3).
      #
      # Byte-faithful port of the verified Rust reference
      # (`ports/rust/src/security/identity/emoji_zwj_integrity.rs`) and of
      # `Unicode/Security/Identity/EmojiZwjIntegrity.lean`.
      #
      # Threat model.  An adversary crafts an emoji-shaped codepoint sequence
      # containing one or more U+200D ZERO WIDTH JOINERs but violating the
      # sanctioned RGI ZWJ-sequence shape — by exceeding the RGI length cap, by
      # joining a non-emoji codepoint, by emitting adjacent ZWJ pairs, or by
      # overflowing the skin-tone count.  Any non-RGI ZWJ-containing sequence is
      # renderer-dependent, and that renderer divergence is the attack surface.
      #
      # Sanctioning data.  UTS #51 defines the RGI ZWJ sequences in
      # `emoji-zwj-sequences.txt`, bundled byte-identically in this port's own
      # `data/emoji-zwj-sequences.txt` and parsed here with the port's own text
      # idiom (never a host emoji library, never String normalization).  The
      # registered set gives both the exact-match membership test
      # (`registered_zwj_sequence?`) and the ZWJ *alphabet* — every distinct
      # codepoint occurring at any position of any registered sequence, excluding
      # the joiner — which is the canonical "what may flank a ZWJ?" predicate.
      #
      # Algorithm (one pass over `input`).
      #   Phase 1 — collect ZWJ positions and the skin-tone count.
      #   Phase 2 — short-circuit Clear if there are no ZWJs and the skin-tone
      #             count is at most 1.
      #   Phase 3 — a registered RGI sequence is always Clear.
      #   Phase 4 — check sub-threats by priority:
      #               1. DoubleZWJ            ZWJ-ZWJ adjacency
      #               2. NonEmojiInjection    ZWJ adjacent to a non-emoji codepoint
      #               3. OverLength           sequence longer than the RGI cap
      #               4. SkinToneOverflow     skin-tone count >= 5
      #               5. UnregisteredSequence catch-all when ZWJs are present but
      #                                       the sequence is not registered.
      module EmojiZwjIntegrity
        # Conservative cap on the length of a sanctioned RGI ZWJ sequence
        # (`maxRgiLength` in the Lean spec).  The longest current entry (a
        # four-person family with skin tones) reaches ~13-14 codepoints; 16 is a
        # safe upper bound.
        MAX_RGI_LENGTH = 16

        # The ZERO WIDTH JOINER codepoint.
        ZWJ = 0x200D

        # A sub-threat this detector can fire.  `tag` is the fixture-row / wire
        # tag string; `data` carries the variant-specific position payload
        # (`positions`, `zwj_pos` + `non_emoji_cp`, `length` + `max_length`,
        # `count`, or `chain_len`).
        SubThreat = Struct.new(:tag, :data)

        # Top-level classification.  `sub` is nil when clear; `positions` is the
        # codepoint indices the sub-threat implicates (empty when clear);
        # `decoded` is the decoded-byte projection (always empty for this
        # detector, kept for shape parity with the Lean `Classification.hazard`).
        Classification = Struct.new(:sub, :positions, :decoded) do
          def clear?
            sub.nil?
          end

          def tag
            sub.nil? ? nil : sub.tag
          end
        end

        # The structured output of `detect` (mirrors the Lean `Verdict`).
        # `chain_length` is 0 when there are no ZWJs, else the input length;
        # `is_registered_rgi` is true iff the input is exactly a registered RGI
        # ZWJ sequence; `skin_tone_count` counts U+1F3FB..U+1F3FF.
        Verdict = Struct.new(
          :input, :classify, :zwj_positions, :chain_length,
          :is_registered_rgi, :skin_tone_count
        )

        module_function

        # ── Sub-threat constructors ────────────────────────────────────────

        # ZWJ-ZWJ adjacency; `positions` are the first ZWJ of each adjacent pair.
        def double_zwj(positions)
          SubThreat.new("DoubleZWJ", { positions: positions })
        end

        # A ZWJ flanked by a non-emoji codepoint (or sitting at an input edge).
        # `non_emoji_cp` is the offending codepoint (0 for an edge ZWJ).
        def non_emoji_injection(zwj_pos, non_emoji_cp)
          SubThreat.new("NonEmojiInjection", { zwj_pos: zwj_pos, non_emoji_cp: non_emoji_cp })
        end

        # The sequence is longer than `MAX_RGI_LENGTH`.
        def over_length(length, max_length)
          SubThreat.new("OverLength", { length: length, max_length: max_length })
        end

        # Five or more skin-tone modifiers (the family-emoji maximum is four).
        def skin_tone_overflow(count)
          SubThreat.new("SkinToneOverflow", { count: count })
        end

        # ZWJs are present and no other sub-threat matched, but the sequence is
        # not a registered RGI ZWJ sequence.
        def unregistered_sequence(chain_len)
          SubThreat.new("UnregisteredSequence", { chain_len: chain_len })
        end

        # ── Classification constructors ────────────────────────────────────

        def clear
          Classification.new(nil, [], [])
        end

        def hazard(sub, positions, decoded)
          Classification.new(sub, positions, decoded)
        end

        # ── RGI ZWJ-sequence data (bundled data/emoji-zwj-sequences.txt) ────

        # Parse a hex codepoint token, or nil when the token is not all-hex.
        def parse_hex(str)
          token = str.strip
          return nil if token.empty?
          return nil unless token.match?(/\A[0-9A-Fa-f]+\z/)

          token.to_i(16)
        end

        # The registered RGI ZWJ sequences.  Each non-comment row is
        # `<cp> <cp> ... ; RGI_Emoji_ZWJ_Sequence ; <desc> # <cmt>`; the
        # codepoint list is the field before the first `;`.
        def zwj_sequences
          @zwj_sequences ||= parse_zwj_sequences
        end

        def parse_zwj_sequences
          out = []
          UnicodeRuby.read_data("emoji-zwj-sequences.txt").each_line do |raw|
            line = raw.chomp
            idx = line.index("#")
            body = idx.nil? ? line : line[0...idx]
            stripped = body.strip
            next if stripped.empty?

            seq_field = stripped.split(";").first
            next if seq_field.nil?

            seq = []
            parsed_ok = true
            seq_field.split.each do |token|
              cp = parse_hex(token)
              if cp.nil?
                parsed_ok = false
                break
              end
              seq << cp
            end
            out << seq if parsed_ok && !seq.empty?
          end
          out
        end

        # The ZWJ alphabet: every distinct codepoint occurring at any position of
        # any registered RGI ZWJ sequence, excluding the joiner U+200D itself.
        def zwj_alphabet
          @zwj_alphabet ||= build_zwj_alphabet
        end

        def build_zwj_alphabet
          set = Set.new
          zwj_sequences.each do |seq|
            seq.each do |cp|
              set.add(cp) if cp != ZWJ
            end
          end
          set
        end

        # True iff `cps` is exactly a registered RGI ZWJ sequence.
        def registered_zwj_sequence?(cps)
          zwj_sequences.any? { |seq| seq == cps }
        end

        # True iff `cp` appears at some position of a registered RGI ZWJ sequence
        # (the canonical "what may flank a ZWJ?" predicate).
        def emoji_target?(cp)
          zwj_alphabet.include?(cp)
        end

        # ── Core predicates ────────────────────────────────────────────────

        # True iff `cp` is the ZWJ codepoint.
        def zwj?(cp)
          cp == ZWJ
        end

        # True iff `cp` is an emoji skin-tone modifier (U+1F3FB..U+1F3FF).
        def emoji_modifier?(cp)
          cp >= 0x1F3FB && cp <= 0x1F3FF
        end

        # Positions of every ZWJ in `input`.
        def zwj_positions(input)
          out = []
          input.each_index { |i| out << i if zwj?(input[i]) }
          out
        end

        # Count of skin-tone modifier codepoints.
        def skin_tone_count(input)
          input.count { |cp| emoji_modifier?(cp) }
        end

        # Positions of the first ZWJ in each ZWJ-ZWJ adjacent pair.
        def double_zwj_positions(input)
          out = []
          (0...input.length).each do |idx|
            next_cp = input[idx + 1]
            unless next_cp.nil?
              out << idx if zwj?(input[idx]) && zwj?(next_cp)
            end
          end
          out
        end

        # The first ZWJ position where either neighbour is a non-emoji codepoint,
        # as `[zwj_pos, offending_cp]`, or nil when none.  A ZWJ at an input edge
        # (no preceding or no following codepoint) is itself an injection-class
        # hazard, reported with offending codepoint 0.
        def first_non_emoji_injection(input)
          (0...input.length).each do |idx|
            next unless zwj?(input[idx])

            prev = idx.zero? ? nil : input[idx - 1]
            nxt = input[idx + 1]
            if !prev.nil? && !nxt.nil?
              if !emoji_target?(prev)
                return [idx, prev]
              elsif !emoji_target?(nxt)
                return [idx, nxt]
              end
            elsif prev.nil?
              return [idx, 0]
            else
              return [idx, 0]
            end
          end
          nil
        end

        # ── Top-level detection ────────────────────────────────────────────

        # The EmojiZwjIntegrity detection function.
        def detect(input)
          zwjs = zwj_positions(input)
          st_count = skin_tone_count(input)
          is_rgi = registered_zwj_sequence?(input)
          chain_len = zwjs.empty? ? 0 : input.length

          if zwjs.empty? && st_count <= 1
            return Verdict.new(input.dup, clear, [], 0, is_rgi, st_count)
          end

          classification =
            if is_rgi
              # Phase 3: a registered RGI sequence is always clear.
              clear
            else
              # Phase 4.1: ZWJ-ZWJ adjacency.
              dzwj = double_zwj_positions(input)
              if !dzwj.empty?
                hazard(double_zwj(dzwj.dup), dzwj, [])
              else
                # Phase 4.2: ZWJ adjacent to a non-emoji codepoint.
                injection = first_non_emoji_injection(input)
                if !injection.nil?
                  zwj_pos, offend_cp = injection
                  hazard(non_emoji_injection(zwj_pos, offend_cp), [zwj_pos], [])
                elsif input.length > MAX_RGI_LENGTH
                  # Phase 4.3: length cap.
                  hazard(over_length(input.length, MAX_RGI_LENGTH), [], [])
                elsif st_count >= 5
                  # Phase 4.4: skin-tone overflow.
                  hazard(skin_tone_overflow(st_count), [], [])
                elsif !zwjs.empty?
                  # Phase 4.5: catch-all for unregistered ZWJ sequences.
                  hazard(unregistered_sequence(input.length), zwjs.dup, [])
                else
                  clear
                end
              end
            end

          Verdict.new(input.dup, classification, zwjs, chain_len, is_rgi, st_count)
        end
      end
    end
  end
end

File: ports/php/src/Security/Identity/EmojiZwjIntegrity.php
<?php

declare(strict_types=1);

// EmojiZwjIntegrity — detection of malformed / unsanctioned emoji ZWJ-sequence
// shapes per UTS #51 (the identity-layer detector I3).
//
// Byte-faithful port of the verified rust reference
// `ports/rust/src/security/identity/emoji_zwj_integrity.rs` and of
// `Unicode/Security/Identity/EmojiZwjIntegrity.lean`.
//
// Threat model. An adversary crafts an emoji-shaped codepoint sequence
// containing one or more U+200D ZERO WIDTH JOINERs but violating the sanctioned
// RGI ZWJ-sequence shape — by exceeding the RGI length cap, by joining a
// non-emoji codepoint, by emitting adjacent ZWJ pairs, or by overflowing the
// skin-tone count. Any non-RGI ZWJ-containing sequence is renderer-dependent,
// and that renderer divergence is the attack surface.
//
// Sanctioning data. UTS #51 defines the RGI ZWJ sequences in
// `emoji-zwj-sequences.txt`, bundled byte-identically in this port's own
// `data/emoji-zwj-sequences.txt` and parsed here with the port's own text idiom
// (never a host emoji library, never String normalization). The registered set
// gives both the exact-match membership test (`isRegisteredZwjSequence`) and the
// ZWJ *alphabet* — every distinct codepoint occurring at any position of any
// registered sequence, excluding the joiner — which is the canonical "what may
// flank a ZWJ?" predicate.
//
// Algorithm (one pass over `input`).
//   Phase 1 — collect ZWJ positions and the skin-tone count.
//   Phase 2 — short-circuit Clear if there are no ZWJs and the skin-tone count
//             is at most 1.
//   Phase 3 — a registered RGI sequence is always Clear.
//   Phase 4 — check sub-threats by priority:
//               1. DoubleZWJ            ZWJ-ZWJ adjacency
//               2. NonEmojiInjection    ZWJ adjacent to a non-emoji codepoint
//               3. OverLength           sequence longer than the RGI cap
//               4. SkinToneOverflow     skin-tone count >= 5
//               5. UnregisteredSequence catch-all when ZWJs are present but the
//                                       sequence is not registered.
//
// PHP has no module system, so the support types the rust reference keeps in a
// dedicated module (SubThreat, Classification, Verdict) are carried here under an
// `EmojiZwj` class-name prefix — the identity namespace already binds bare names
// for the sibling homoglyph detector's helpers. The tag strings, variant order,
// and priority all mirror the reference exactly.

namespace UnicodePhp\Security\Identity;

use UnicodePhp\Data;

// ─────────────────────────────────────────────────────────────────────
// §1 Types
// ─────────────────────────────────────────────────────────────────────

/**
 * A sub-threat this detector can fire, in priority order. Each variant carries
 * only the payload the rust reference records for it; the unused payload fields
 * are null. The tag is the human-facing classification string, mirrored verbatim
 * from the rust reference.
 */
final class EmojiZwjSubThreat
{
    /**
     * @param list<int>|null $positions First ZWJ of each ZWJ-ZWJ pair (DoubleZWJ only).
     */
    private function __construct(
        public readonly string $kind,
        public readonly ?array $positions,
        public readonly ?int $zwjPos,
        public readonly ?int $nonEmojiCp,
        public readonly ?int $length,
        public readonly ?int $maxLength,
        public readonly ?int $count,
        public readonly ?int $chainLen,
    ) {
    }

    /**
     * ZWJ-ZWJ adjacency; $positions are the first ZWJ of each adjacent pair.
     *
     * @param list<int> $positions
     */
    public static function doubleZwj(array $positions): EmojiZwjSubThreat
    {
        return new EmojiZwjSubThreat('DoubleZWJ', $positions, null, null, null, null, null, null);
    }

    /**
     * A ZWJ flanked by a non-emoji codepoint (or sitting at an input edge).
     * $zwjPos is the offending ZWJ; $nonEmojiCp is the flanking codepoint (0 for
     * an edge ZWJ).
     */
    public static function nonEmojiInjection(int $zwjPos, int $nonEmojiCp): EmojiZwjSubThreat
    {
        return new EmojiZwjSubThreat('NonEmojiInjection', null, $zwjPos, $nonEmojiCp, null, null, null, null);
    }

    /** The sequence is longer than the RGI cap; $length observed, $maxLength the cap. */
    public static function overLength(int $length, int $maxLength): EmojiZwjSubThreat
    {
        return new EmojiZwjSubThreat('OverLength', null, null, null, $length, $maxLength, null, null);
    }

    /** Five or more skin-tone modifiers (the family-emoji maximum is four); $count observed. */
    public static function skinToneOverflow(int $count): EmojiZwjSubThreat
    {
        return new EmojiZwjSubThreat('SkinToneOverflow', null, null, null, null, null, $count, null);
    }

    /**
     * ZWJs are present and no other sub-threat matched, but the sequence is not a
     * registered RGI ZWJ sequence; $chainLen is the length of the unregistered chain.
     */
    public static function unregisteredSequence(int $chainLen): EmojiZwjSubThreat
    {
        return new EmojiZwjSubThreat('UnregisteredSequence', null, null, null, null, null, null, $chainLen);
    }

    /** Fixture-row / wire tag string for this sub-threat (matches `SubThreat.tag`). */
    public function tag(): string
    {
        return match ($this->kind) {
            'DoubleZWJ' => 'DoubleZWJ',
            'NonEmojiInjection' => 'NonEmojiInjection',
            'OverLength' => 'OverLength',
            'SkinToneOverflow' => 'SkinToneOverflow',
            'UnregisteredSequence' => 'UnregisteredSequence',
            default => throw new \RuntimeException("EmojiZwjSubThreat: unknown kind '{$this->kind}'"),
        };
    }
}

/**
 * Top-level classification for EmojiZwjIntegrity: Clear, or a Hazard carrying the
 * fired sub-threat, the implicated positions, and the (always-empty for this
 * detector) decoded-byte projection kept for shape parity with the Lean
 * `Classification.hazard`.
 */
final class EmojiZwjClassification
{
    /**
     * @param list<int> $positions
     * @param list<int> $decoded
     */
    private function __construct(
        public readonly ?EmojiZwjSubThreat $sub,
        public readonly array $positions,
        public readonly array $decoded,
    ) {
    }

    public static function clear(): EmojiZwjClassification
    {
        return new EmojiZwjClassification(null, [], []);
    }

    /**
     * @param list<int> $positions
     * @param list<int> $decoded
     */
    public static function hazard(EmojiZwjSubThreat $sub, array $positions, array $decoded): EmojiZwjClassification
    {
        return new EmojiZwjClassification($sub, $positions, $decoded);
    }

    /** True iff the classification is Clear. */
    public function isClear(): bool
    {
        return $this->sub === null;
    }

    /** Human-facing tag for a hazard, or null when clear. */
    public function tag(): ?string
    {
        return $this->sub?->tag();
    }

    /** @return list<int> Implicated positions (empty when clear). */
    public function positions(): array
    {
        return $this->positions;
    }
}

/**
 * The structured output of `detect` (mirrors the Lean `Verdict`).
 * $chainLength is 0 when there are no ZWJs, else the input length;
 * $isRegisteredRgi is true iff the input is exactly a registered RGI ZWJ
 * sequence; $skinToneCount counts U+1F3FB..U+1F3FF.
 */
final class EmojiZwjVerdict
{
    /**
     * @param list<int> $input
     * @param list<int> $zwjPositions
     */
    public function __construct(
        public readonly array $input,
        public readonly EmojiZwjClassification $classify,
        public readonly array $zwjPositions,
        public readonly int $chainLength,
        public readonly bool $isRegisteredRgi,
        public readonly int $skinToneCount,
    ) {
    }
}

final class EmojiZwjIntegrity
{
    // ─────────────────────────────────────────────────────────────────
    // §1 Constants
    // ─────────────────────────────────────────────────────────────────

    /**
     * Conservative cap on the length of a sanctioned RGI ZWJ sequence
     * (`maxRgiLength` in the Lean spec). The longest current entry (a four-person
     * family with skin tones) reaches ~13-14 codepoints; 16 is a safe upper bound.
     */
    public const MAX_RGI_LENGTH = 16;

    /** The ZERO WIDTH JOINER codepoint. */
    public const ZWJ = 0x200D;

    /** @var list<list<int>>|null Cached registered RGI ZWJ sequences. */
    private static ?array $zwjSequences = null;

    /** @var array<int,bool>|null Cached ZWJ alphabet, keyed by codepoint. */
    private static ?array $zwjAlphabet = null;

    // ─────────────────────────────────────────────────────────────────
    // §2 RGI ZWJ-sequence data (bundled data/emoji-zwj-sequences.txt)
    // ─────────────────────────────────────────────────────────────────

    /**
     * Parse the registered RGI ZWJ sequences from `emoji-zwj-sequences.txt`. Each
     * non-comment row is `<cp> <cp> ... ; RGI_Emoji_ZWJ_Sequence ; <desc> # <cmt>`;
     * the codepoint list is the whitespace-separated field before the first `;`.
     *
     * @return list<list<int>>
     */
    private static function zwjSequences(): array
    {
        if (self::$zwjSequences !== null) {
            return self::$zwjSequences;
        }
        $out = [];
        foreach (Data::lines('emoji-zwj-sequences.txt') as $rawLine) {
            $hash = strpos($rawLine, '#');
            $body = $hash === false ? $rawLine : substr($rawLine, 0, $hash);
            $stripped = trim($body);
            if ($stripped === '') {
                continue;
            }
            $semicolon = strpos($stripped, ';');
            $seqField = $semicolon === false ? $stripped : substr($stripped, 0, $semicolon);
            $seq = [];
            $parsedOk = true;
            foreach (preg_split('/\s+/', trim($seqField), -1, PREG_SPLIT_NO_EMPTY) as $token) {
                $cp = self::parseHex($token);
                if ($cp === null) {
                    $parsedOk = false;
                    break;
                }
                $seq[] = $cp;
            }
            if ($parsedOk && $seq !== []) {
                $out[] = $seq;
            }
        }
        self::$zwjSequences = $out;
        return $out;
    }

    /** Parse a trimmed hexadecimal codepoint token, or null if not all-hex. */
    private static function parseHex(string $s): ?int
    {
        $s = trim($s);
        if ($s === '' || !ctype_xdigit($s)) {
            return null;
        }
        return intval($s, 16);
    }

    /**
     * The ZWJ alphabet: every distinct codepoint occurring at any position of any
     * registered RGI ZWJ sequence, excluding the joiner U+200D itself.
     *
     * @return array<int,bool>
     */
    private static function zwjAlphabet(): array
    {
        if (self::$zwjAlphabet !== null) {
            return self::$zwjAlphabet;
        }
        $set = [];
        foreach (self::zwjSequences() as $seq) {
            foreach ($seq as $cp) {
                if ($cp !== self::ZWJ) {
                    $set[$cp] = true;
                }
            }
        }
        self::$zwjAlphabet = $set;
        return $set;
    }

    /**
     * True iff $cps is exactly a registered RGI ZWJ sequence.
     *
     * @param list<int> $cps
     */
    public static function isRegisteredZwjSequence(array $cps): bool
    {
        foreach (self::zwjSequences() as $seq) {
            if ($seq === $cps) {
                return true;
            }
        }
        return false;
    }

    /**
     * True iff $cp appears at some position of a registered RGI ZWJ sequence
     * (the canonical "what may flank a ZWJ?" predicate).
     */
    public static function isEmojiTarget(int $cp): bool
    {
        return isset(self::zwjAlphabet()[$cp]);
    }

    // ─────────────────────────────────────────────────────────────────
    // §3 Core predicates
    // ─────────────────────────────────────────────────────────────────

    /** True iff $cp is the ZWJ codepoint. */
    public static function isZwj(int $cp): bool
    {
        return $cp === self::ZWJ;
    }

    /** True iff $cp is an emoji skin-tone modifier (U+1F3FB..U+1F3FF). */
    public static function isEmojiModifier(int $cp): bool
    {
        return $cp >= 0x1F3FB && $cp <= 0x1F3FF;
    }

    /**
     * Positions of every ZWJ in $input.
     *
     * @param list<int> $input
     * @return list<int>
     */
    private static function zwjPositions(array $input): array
    {
        $out = [];
        foreach ($input as $idx => $cp) {
            if (self::isZwj($cp)) {
                $out[] = $idx;
            }
        }
        return $out;
    }

    /**
     * Count of skin-tone modifier codepoints.
     *
     * @param list<int> $input
     */
    private static function skinToneCount(array $input): int
    {
        $count = 0;
        foreach ($input as $cp) {
            if (self::isEmojiModifier($cp)) {
                $count++;
            }
        }
        return $count;
    }

    /**
     * Positions of the first ZWJ in each ZWJ-ZWJ adjacent pair.
     *
     * @param list<int> $input
     * @return list<int>
     */
    private static function doubleZwjPositions(array $input): array
    {
        $out = [];
        $len = count($input);
        for ($idx = 0; $idx < $len; $idx++) {
            if ($idx + 1 >= $len) {
                continue;
            }
            if (self::isZwj($input[$idx]) && self::isZwj($input[$idx + 1])) {
                $out[] = $idx;
            }
        }
        return $out;
    }

    /**
     * The first ZWJ position where either neighbour is a non-emoji codepoint, as
     * `[$zwjPos, $offendingCp]`, or null when none. A ZWJ at an input edge (no
     * preceding or no following codepoint) is itself an injection-class hazard,
     * reported with offending codepoint 0.
     *
     * @param list<int> $input
     * @return array{0:int,1:int}|null
     */
    private static function firstNonEmojiInjection(array $input): ?array
    {
        $len = count($input);
        for ($idx = 0; $idx < $len; $idx++) {
            if (!self::isZwj($input[$idx])) {
                continue;
            }
            $prev = $idx === 0 ? null : $input[$idx - 1];
            $next = $idx + 1 < $len ? $input[$idx + 1] : null;
            if ($prev !== null && $next !== null) {
                if (!self::isEmojiTarget($prev)) {
                    return [$idx, $prev];
                } elseif (!self::isEmojiTarget($next)) {
                    return [$idx, $next];
                }
            } elseif ($prev === null) {
                return [$idx, 0];
            } else {
                return [$idx, 0];
            }
        }
        return null;
    }

    // ─────────────────────────────────────────────────────────────────
    // §4 Top-level detection
    // ─────────────────────────────────────────────────────────────────

    /**
     * The EmojiZwjIntegrity detection function. Runs the priority ladder over the
     * one-pass scan results; the first sub-threat wins. See the file header for
     * the ordering rationale.
     *
     * @param list<int> $input
     */
    public static function detect(array $input): EmojiZwjVerdict
    {
        $input = array_values($input);

        $zwjs = self::zwjPositions($input);
        $stCount = self::skinToneCount($input);
        $isRgi = self::isRegisteredZwjSequence($input);
        $chainLen = $zwjs === [] ? 0 : count($input);

        if ($zwjs === [] && $stCount <= 1) {
            return new EmojiZwjVerdict($input, EmojiZwjClassification::clear(), [], 0, $isRgi, $stCount);
        }

        if ($isRgi) {
            // Phase 3: a registered RGI sequence is always clear.
            $classification = EmojiZwjClassification::clear();
        } else {
            // Phase 4.1: ZWJ-ZWJ adjacency.
            $dzwj = self::doubleZwjPositions($input);
            if ($dzwj !== []) {
                $classification = EmojiZwjClassification::hazard(
                    EmojiZwjSubThreat::doubleZwj($dzwj),
                    $dzwj,
                    [],
                );
            } else {
                // Phase 4.2: ZWJ adjacent to a non-emoji codepoint.
                $injection = self::firstNonEmojiInjection($input);
                if ($injection !== null) {
                    [$zwjPos, $offendCp] = $injection;
                    $classification = EmojiZwjClassification::hazard(
                        EmojiZwjSubThreat::nonEmojiInjection($zwjPos, $offendCp),
                        [$zwjPos],
                        [],
                    );
                } elseif (count($input) > self::MAX_RGI_LENGTH) {
                    // Phase 4.3: length cap.
                    $classification = EmojiZwjClassification::hazard(
                        EmojiZwjSubThreat::overLength(count($input), self::MAX_RGI_LENGTH),
                        [],
                        [],
                    );
                } elseif ($stCount >= 5) {
                    // Phase 4.4: skin-tone overflow.
                    $classification = EmojiZwjClassification::hazard(
                        EmojiZwjSubThreat::skinToneOverflow($stCount),
                        [],
                        [],
                    );
                } elseif ($zwjs !== []) {
                    // Phase 4.5: catch-all for unregistered ZWJ sequences.
                    $classification = EmojiZwjClassification::hazard(
                        EmojiZwjSubThreat::unregisteredSequence(count($input)),
                        $zwjs,
                        [],
                    );
                } else {
                    $classification = EmojiZwjClassification::clear();
                }
            }
        }

        return new EmojiZwjVerdict($input, $classification, $zwjs, $chainLen, $isRgi, $stCount);
    }
}
